[FLINK-28419][table] Add runtime provider interface for full caching lookup + implement Periodic and Timed cache reload triggers

SmirAlex opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request introduces new interfaces that allows to use Full cache as `LookupFunctionProvider`. Interface list:

- `FullCachingLookupProvider`
- `CacheReloadTrigger`

CacheReloadTrigger provides customizable way when to trigger cache reload operation. There are 2 built-in triggers:

- `PeriodicCacheReloadTrigger` - cache is reloaded with fixed intervals without initial delay
- `TimedCacheReloadTrigger` - cache is reloaded at specified time with fixed intervals multiples of one day

Also there are introduced common options for connectors that will support Full caching lookup join.


## Brief change log

- Add interfaces FullCachingLookupProvider and CacheReloadTrigger
- Implement reload triggers PeriodicCacheReloadTrigger and TimedCacheReloadTrigger
- Add common SQL options for Full caching lookup join


## Verifying this change

This change added unit tests for the newly added `PeriodicCacheReloadTrigger` and `TimedCacheReloadTrigger`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (FLIP docs + JavaDocs)

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/CacheReloadTrigger.java:
```diff
@@ -0,0 +1,51 @@
+     * Context of {@link CacheReloadTrigger} for getting information about times and triggering
+     * reload.
+     */
+    interface Context {
```
Add @PublicEvolving here

SmirAlex replied:
Fixed

PatrickRen reviewed (commented):
@SmirAlex Thanks for the contribution! The PR overall looks good to me. Just some minor comments.

SmirAlex commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTriggerTest.java:
```diff
@@ -0,0 +1,141 @@
+    private final ScheduleStrategyExecutorService scheduledExecutor =
+            new ScheduleStrategyExecutorService();
+
+    private final Runnable reloadTask = Mockito.mock(Runnable.class);
```
Yea, I just thought this little one would be ok, because basically we don't need any implementation of this Runnable, except counting calls of one method. Anyway I fixed that, no problem.

SmirAlex commented:
@PatrickRen  Thanks for the review! I've made fixes according to your comments. Please look at the latest commit, when you will have time.

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java:
```diff
@@ -0,0 +1,131 @@
+ * ScheduleMode}. If {@code reloadInterval} is zero, cache loading will happen just one time.
+ */
+public class PeriodicCacheReloadTrigger implements CacheReloadTrigger {
+    private static final long serialVersionUID = 3828732577291369913L;
```
The serial version UID should start from 1 for new classes

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public void close() throws Exception {
+        scheduledExecutor.shutdownNow();
```
What about wrapping with a `if(scheduledExecutor != null)`?

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/TimedCacheReloadTrigger.java:
```diff
@@ -0,0 +1,141 @@
+
+/** A trigger that reloads at a specific time and repeats for the given interval in days. */
+public class TimedCacheReloadTrigger implements CacheReloadTrigger {
+    private static final long serialVersionUID = -8565574480311775185L;
```
The serial version UID should start from 1 for new classes

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/TimedCacheReloadTrigger.java:
```diff
@@ -0,0 +1,141 @@
+
+    @Override
+    public void close() throws Exception {
+        scheduledExecutor.shutdownNow();
```
What about wrapping with a if(scheduledExecutor != null)?

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/LookupOptions.java:
```diff
@@ -0,0 +1,93 @@
+    public static final ConfigOption<Duration> FULL_CACHE_PERIODIC_RELOAD_INTERVAL =
+            ConfigOptions.key("lookup.full-cache.periodic-reload.interval")
+                    .durationType()
+                    .defaultValue(Duration.ZERO)
```
I'm not sure if it is a good idea to have a default value with 0 here. If users intend to reload the cache periodically, the interval should be set in the configuration to fully express the reloading strategy. Maybe we could just leave it without default value here and have a check in the `PeriodicCacheReloadTrigger`. WDYT?

SmirAlex replied:
Yes, I agree. If partial cache can't be built without specifying options for it, the same should be for full cache.

PatrickRen reviewed (commented):
Thanks for the update @SmirAlex ! I left some comments. Also please rebase the latest master as the `LookupOptions` has been merged to master.

SmirAlex commented:
@PatrickRen I resolved your comments and rebased. Additionally I changed error messages and config descriptions, so they now correspond to the ones for partial caching. You can have a look. Do I need to run CI again before merge?

PatrickRen commented:
@SmirAlex Yeah we have to wait for a green CI run. It should be running now so let's wait for a while.

SmirAlex commented:
@flinkbot run azure

SmirAlex commented:
@PatrickRen There were multiple failed pipelines, but they were because of unrelated issues. Latest pipeline was successful - https://dev.azure.com/asmirnov1/Flink/_build/results?buildId=17&view=results

PatrickRen reviewed (approved):
Thanks for the contribution @SmirAlex ! LGTM

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/FullCachingLookupProvider.java:
```diff
@@ -0,0 +1,63 @@
+ */
+@PublicEvolving
+public interface FullCachingLookupProvider extends LookupFunctionProvider {
+    static FullCachingLookupProvider of(
```
Missing JavaDoc for this public API method

PatrickRen commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/lookup/cache/trigger/PeriodicCacheReloadTrigger.java:
```diff
@@ -0,0 +1,135 @@
+
+/**
+ * A trigger that reloads cache entries periodically with specified interval and {@link
+ * ScheduleMode}. If {@code reloadInterval} is zero, cache loading will happen just one time.
```
I think this JavaDoc should be updated as the argument check in the constructor has disabled the possibility of using zero as the reload interval.

PatrickRen reviewed (commented):
Just two minor comments about JavaDoc.

SmirAlex commented:
@PatrickRen thanks for the review! I've fixed last comments about javadoc.

PatrickRen commented:
Merged on master 5405239dec0884dff746129c73955c90f455c465

xiaolan-bit commented:
I would like to ask, as I have a scenario where Flink writes into Hudi. Due to the complexity of the SQL process, I want to turn the whole process into an SQL file. However, due to the latency of the insert operation, I want to delay the operation through Flink SQL statements. Is there such a statement that can implement the delay operation, for example, execute the next SQL statement only after 5 seconds?
